File: src/emulator/fetch.ts
import { ok, Result } from "types/result";
import { InstructionSet } from "./instruction_set";
import { CoreState } from "./state";

export type Argument = ByteArgument | WordArgument | AddressArgument | MaxArgument | RegisterArgument | PortArgument;
export type ByteArgument = { type: "word", value: number };
export type WordArgument = { type: "word", value: number };
export type AddressArgument = { type: "address", value: number };
export type MaxArgument = { type: "max", value: number };
export type RegisterArgument = { type: "register", value: number };
export type PortArgument = { type: "port", value: number };
type Instruction = {
  name: string;
  arguments: Argument[];
}

export function fetch(core: CoreState): Result<Instruction, string> {
  const instruction_head = core.memoryctl.getInstruction(core.registers.address[0]);
  if (instruction_head.is_err()) return instruction_head;
  core.registers.address[0] += core.memoryctl.bitwidths.byte == 1 ? 2 : 1;
  const instruction_structure = InstructionSet[instruction_head.unwrap()];

  const fetched_instruction: Instruction = {
    name: instruction_structure.name,
    arguments: []
  };

  for (let i = 0; i < instruction_structure.argument_count; i++) {
    const type = instruction_structure.argument_types[i];
    switch (type.type) {
      case "port":
      case "register":
      case "word": {
        const v = core.memoryctl.getWord(core.registers.address[0]);
        if (v.is_err()) return v;
        fetched_instruction.arguments.push({
          type: type.type, value: v.unwrap()
        });
        core.registers.address[0] += core.memoryctl.bitwidths.word;
        break;
      };
      case "max":
      case "address": {
        const v = core.memoryctl.getAddress(core.registers.address[0]);
        if (v.is_err()) return v;
        fetched_instruction.arguments.push({
          type: type.type, value: v.unwrap()
        });
        core.registers.address[0] += core.memoryctl.bitwidths.address;
        break;
      }
    }
  }

  return ok(fetched_instruction);
}

File: src/emulator/fetch.zig
const std = @import("std");
const emulator = @import("./state.zig");
const instructions = @import("./instructions.zig");

const allocator = std.heap.page_allocator;
const InCode = instructions.InCode;
const Core = emulator.Core;

fn Instruction(comptime WordType: type, comptime AddrType: type) type {
    return struct {
        name: InCode,
        arguments: []InstructionArgument(WordType, AddrType)
    };
}

fn InstructionArgument(comptime WordType: type, comptime AddrType: type) type {
    return union(instructions.InstructionArgumentType) {
        REGISTER: struct {
            value: WordType
        },
        WORD: struct {
            value: WordType
        },
        ADDRESS: struct {
            value: AddrType
        },
        PORT: struct {
            value: WordType
        },
    };
}

pub fn fetch(comptime ByteType: type, comptime WordType: type, comptime AddrType: type, core: *Core(ByteType, WordType, AddrType)) !*Instruction(WordType, AddrType) {
    const byteBitwidth = @bitSizeOf(ByteType);
    const instructionLength = instructions.Instructions.len;
    const toFloor: f64= @floatFromInt(instructionLength);
    const opcodeBitwidth: i32 = @intFromFloat(std.math.floor(std.math.log2(toFloor)) + 1.0);
    const byteAmount: usize = @intCast(try std.math.divCeil(i32, opcodeBitwidth, byteBitwidth));

    const pc = core.registers[0];

    var opcode: usize = 0;
    for (0..byteAmount) |_| {
        opcode <<= byteBitwidth;
        opcode |= core.memoryController.fetchByte(pc.Addr.value); 
        pc.Addr.value += 1;
    }

    const instr = instructions.Instructions[opcode];
    
    const instruction: *Instruction(WordType, AddrType) = try allocator.create(Instruction(WordType, AddrType));
    instruction.name = instr.name;
    instruction.arguments = try allocator.alloc(InstructionArgument(WordType, AddrType), instr.arguments.len);
    
    for (instr.arguments, 0..) |argument, i| {
        switch (argument) {
            .PORT => {
                const res = try core.memoryController.fetchWord(pc.Addr.value);
                pc.Addr.value += @intCast(res.bytes);
                instruction.arguments[i] 
                    = InstructionArgument(WordType, AddrType){ .PORT = .{
                            .value = res.result
                        }
                    };
            },
            .WORD => {
                const res = try core.memoryController.fetchWord(pc.Addr.value);
                pc.Addr.value += @intCast(res.bytes);
                instruction.arguments[i] = InstructionArgument(WordType, AddrType) {.WORD = .{
                    .value = res.result
                }
    };
            },
            .REGISTER => {
                const res = try core.memoryController.fetchWord(pc.Addr.value);
                pc.Addr.value += @intCast(res.bytes);
                instruction.arguments[i] = InstructionArgument(WordType, AddrType) {.REGISTER = .{
                    .value = res.result
                }};
            },
            .ADDRESS => {
                const res = try core.memoryController.fetchAddr(pc.Addr.value);
                pc.Addr.value += @intCast(res.bytes);
                instruction.arguments[i] = InstructionArgument(WordType, AddrType){.ADDRESS = .{ 
                    .value = res.result 
                }};
            }
        }
    }

    return instruction;
}
